029-across-step/proposal.md

```diff
@@ -0,0 +1,199 @@
+# `across` step
+
+This proposal introduces a mechanism by which a build plan can be executed
+multiple times with different `((vars))`, in service of build matxies, pipeline
+matrixes, and - building on [`((var))` sources][var-sources-rfc] - dynamic
+variations of both.
+
+
+## Motivation
+
+* Support dynamic multi-branch workflows:
+  [concourse/concourse#1172][multi-branch-issue]
+* Support build matrixes and pipeline matrixes.
+* Replace a common, and very flawed, usage pattern of `version: every`, which
+  has proven to be very complicated to support.
+
+
+## Proposal
+
+The `across` step modifier is given a list containing vars and associated
+values to set when executing the step. The step will be executed across all
+combinations of var values.
+
+### Static var values
+
+The set of values can be static, as below:
+
+```yaml
+task: unit
+vars: {go_version: ((.:go_version))}
+across:
+- var: go_version
+  values: [1.12, 1.13]
+```
+
+This will run the `unit` task twice: once with `go_version` set to `1.12`, and
+again with `1.13`.
+
+### Dynamic var values from a var source
+
+Rather than static values, var values may be pulled from a var source
+dynamically:
+
+```yaml
+var_sources:
+- name: booklit-prs
+  type: github-prs
+  config:
+    repository: vito/booklit
+    access_token: # ...
+
+plan:
+- set_pipeline: pr
+  instance_vars: {pr_number: ((.:pr.number))}
+  across:
+  - var: pr
+    source: booklit-prs
+```
+
+The above example will run the `set_pipeline` step across the set of all GitHub
+PRs, returned through a `list` operation on the `booklit-prs` var source.
+
+### Running across a matrix of var values
+
+Multiple vars may be listed to form a matrix:
+
+```yaml
+set_pipeline: pr-go
+instance_vars:
+  pr_number: ((.:pr.number))
+  go_version: ((.:go_version))
+across:
+- var: pr
+  source: booklit-prs
+- var: go_version
+  values: [1.12, 1.13]
+```
+
+This will run 2 * (# of PRs) `set_pipeline` steps, setting two pipelines per
+PR: one for Go 1.12, and one for Go 1.13.
+
+### Controlling parallelism with `max_in_flight`
```

Comment:
The `in_parallel` step also has a `fail_fast` option - is this something that also makes sense for the across step?

Similarly, does it make sense to use the term `limit` rather than `max_in_flight` for consistency with `in_parallel`, or do you think that'd just lead to confusion?

Reply:
Good points!

I had the same debate over `max_in_flight` vs `limit`, but I think `max_in_flight` is less ambiguous in this specific context as  `limit` could also mean "limit to first N vars".

The `in_parallel` step is a bit inconsistent in other ways: we opted to go with a nested config format for its additional settings (`in_parallel: {steps: [...], limit: 3, fail_fast: true}`) to avoid polluting the `PlanConfig` type, but with the step refactor (https://github.com/concourse/concourse/pull/5504) it now has its own `InParallelStep` type, so that wouldn't be a problem anymore, and we could easily change it to this syntax:

```yaml
in_parallel:
- step1
- step2
fail_fast: true
limit: 3
```

I recall @itsdalmo (who PR'd the `in_parallel` step) suggesting it would be confusing to have the term `max_in_flight` in multiple places (on the [job](https://concourse-ci.org/jobs.html#schema.job.max_in_flight) and on the `in_parallel` step), but, well, they all mean the same thing, so it could end up helping folks understand and discover the option. So now I wonder if we should change the `in_parallel` step to `max_in_flight` and a flat config syntax. (It would be easy to do backwards-compatibly; just add more fields to the `InParallelStep` struct.)

Re: `fail_fast` - hmm, I think this definitely makes sense to add, but it would be hard to add in the current syntax, since it'd be an attribute of the whole `across` step modifier, and modifiers can't have additional sibling fields. I was able to get away with putting `max_in_flight` under the vars list but that trick wouldn't make sense for `fail_fast`.

I wonder if I should just change this back to being its own step rather than a step modifier. That would clear up precedence questions between `across` and `timeout`/`attempts`/`on_*`/`ensure` (i.e. whether they apply to the modified step or the original step).

Something like this:

```yaml
across:
- var: go_version
  values: [1.14, 1.15]
- var: postgres_version
  values: [9.5, 10]
each:
  task: unit
  vars:
    go_version: ((.:go_version))
    postgres_version: ((.:postgres_version))
fail_fast: true
```

The only part of this syntax I don't really like is that it raises a question of whether `each` is a list of steps or a single step. Making it a list would make it easier to run multiple steps, but it would be inconsistent with the `on_*` hooks. Keeping it a single step would be consistent with the hooks, but then you'd have to use `do:`. 🤷 

One alternative would be to do something like what `in_parallel` did, which is to support either a list of vars or a config, but, well, it feels kind of clunky:

```yaml
task: unit
vars:
  go_version: ((.:go_version))
  postgres_version: ((.:postgres_version))
across:
  vars:
  - var: go_version
    values: [1.14, 1.15]
  - var: postgres_version
    values: [9.5, 10]
  fail_fast: true
```

Reply:
If we decide to go back to `across` being its own step, what do you think of renaming `each` to `do` (similar to the previous revision of the RFC):

```yaml
across:
- var: go_version
  values: [1.14, 1.15]
- var: postgres_version
  values: [9.5, 10]
do:
- task: unit
  vars:
    go_version: ((.:go_version))
    postgres_version: ((.:postgres_version))
fail_fast: true
```

I think this would make it clear(er) that it's a list of steps rather than a single step (since it looks like a modifier that only applies to the `do` step - although maybe this could lead to some confusion where people would try to apply it to other steps as well?)

Reply:
@aoldershaw Yeah, that's the tricky part: if we overload `do:` it could end up confusing folks into attaching it to other step types, and it might be weird to see a "normal" `do:` after that.

Reply:
For us Elm fans... :P

```yaml
let:
- var: go_version
  values: [1.14, 1.15]
- var: postgres_version
  values: [9.5, 10]
in:
  task: unit
  vars:
    go_version: ((.:go_version))
    postgres_version: ((.:postgres_version))
fail_fast: true
```

Reply:
Hmm...

Here's a thought: if we make it possible for step modifiers to have sibling fields, we could support `fail_fast` in the modifier form and just use the `do:` step as a work-around when it's necessary to disambiguate (pretending `in_parallel` has it as a toplevel field for a moment).

And then if you just put the `do:` below the `across:`, it's exactly the syntax you proposed, but with no gotchas - it's actually a `do` step, and you could replace it with a single step if you wanted. Here's a full example showing its use for disambiguation:

```yaml
across:
- var: go_version
- var: postgres_version
do:
- in_parallel:
  - task: unit
    vars:
      go_version: ((.:go_version))
      postgres_version: ((.:postgres_version))
    on_failure: # send test diagnostics
  fail_fast: false # (or omitted)
on_failure: # send notification
fail_fast: true
```

Precedence would be `ensure` < `on_*` < `attempts` < `timeout` < `across` < (core steps). This would enable modifiers to be applied to the `across` step itself, with the `do:` syntax available for when you need to apply them (or conflicting fields like `fail_fast`) to the inner step.

The cool thing is that this syntax would transparently work while implemented as a modifier - it's just that you're modifying the `do:` step, and you've placed it below its modifier!

It might not be too difficult to support sibling fields - I think it'd just take a bit of `reflect` magic to remove all struct fields instead of just the one used to identify the step. 🤔

Reply:
Huh, that's pretty neat! FWIW I find the `across` syntax a bit easier to follow when written like your example above (with `across` as the first field in the step) rather than when written at the end like other modifiers

Reply:
I can see this possibly being a bit of a gotcha, but what if we didn't remove the sibling `fail_fast` field for `across`, so writing:

```yaml
across:
- var: go_version
- var: postgres_version
in_parallel:
  - task: unit
    vars:
      go_version: ((.:go_version))
      postgres_version: ((.:postgres_version))
  - ...
fail_fast: true
```

would apply the `fail_fast` both to the `across` step **and** to the `in_parallel` step (the assumption being that if you wanted to fail fast, you'd usually want to fail fast everywhere - which may be a wrong assumption). In the case where you didn't want the `across` step or the `in_parallel` step to fail fast, you'd wrap the `in_parallel` in a `do`

Reply:
I've updated the proposal to include `fail_fast` and document the step precedence plus the `do:` syntax workaround.

I ended up going with this precedence order instead: `ensure` < `on_*` < `across` < `attempts` < `timeout`

The reason for this is so that `attempts` is attached to the inner step and not the entire matrix. It doesn't seem like anyone would ever want to rerun the entire matrix in the event that one step fails.

`timeout` is also important to have applied to the inner step: if it were applied to the matrix, `max_in_flight` throttling could easily cause the timeout to be exceeded before some steps even run.